File: iOS/ssabree/ssabree/Features/Board/BoardSearchViewModel.swift
import Foundation
import Observation

// MARK: - UI State

struct BoardSearchUiState {
    var query: String = ""
    var recentKeywords: [String] = []
    var results: [PostModel] = []
    var nextCursor: String? = nil
    var hasNext: Bool = false
    var isLoading: Bool = false
    var isLoadingMore: Bool = false
    var error: String? = nil
    var hasSearched: Bool = false  // 검색을 한 번이라도 했는지
}

// MARK: - ViewModel

@Observable
@MainActor
final class BoardSearchViewModel {
    private let postRepository: PostRepository
    private let keywordRepository: KeywordRepository
    private let pageSize = 20

    var uiState = BoardSearchUiState()

    init(postRepository: PostRepository, keywordRepository: KeywordRepository) {
        self.postRepository = postRepository
        self.keywordRepository = keywordRepository
        loadRecentKeywords()
    }

    // MARK: - Query Change

    func onQueryChange(_ text: String) {
        uiState.query = text
    }

    // MARK: - Search

    func onSearchSubmit() {
        let keyword = uiState.query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !keyword.isEmpty else { return }
        performSearch(keyword: keyword)
    }

    func onQuickSearch(keyword: String) {
        uiState.query = keyword
        performSearch(keyword: keyword)
    }

    private func performSearch(keyword: String) {
        keywordRepository.addRecentKeyword(keyword: keyword)
        loadRecentKeywords()

        Task {
            await load(keyword: keyword, cursor: nil, append: false)
        }
    }

    // MARK: - Load

    private func load(keyword: String, cursor: String?, append: Bool) async {
        if append {
            uiState.isLoadingMore = true
        } else {
            uiState.isLoading = true
            uiState.error = nil
            uiState.hasSearched = true
        }

        let result = await postRepository.getPosts(
            boardId: nil,
            keyword: keyword,
            cursor: cursor,
            limit: pageSize
        )

        switch result {
        case .success(let page):
            if append {
                uiState.results += page.data
            } else {
                uiState.results = page.data
            }
            uiState.nextCursor = page.nextCursor
            uiState.hasNext = page.hasNext
        case .failure(let error):
            uiState.error = error.localizedDescription
        }

        uiState.isLoading = false
        uiState.isLoadingMore = false
    }

    // MARK: - Load More (Infinite Scroll)

    func loadMoreIfNeeded(currentIndex: Int) {
        // 마지막 3개 아이템에 도달하면 더 로드
        let threshold = uiState.results.count - 3
        guard currentIndex >= threshold,
              uiState.hasNext,
              !uiState.isLoadingMore,
              !uiState.isLoading,
              !uiState.query.isEmpty else { return }

        Task {
            await load(keyword: uiState.query, cursor: uiState.nextCursor, append: true)
        }
    }

    // MARK: - Recent Keywords

    private func loadRecentKeywords() {
        uiState.recentKeywords = keywordRepository.getRecentKeywords()
    }

    func deleteRecentKeyword(_ keyword: String) {
        keywordRepository.deleteRecentKeyword(keyword: keyword)
        loadRecentKeywords()
    }

    // MARK: - Clear

    func clearQuery() {
        uiState.query = ""
    }

    func clearError() {
        uiState.error = nil
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/board/screen/BoardSearchViewModel.kt
package com.ssafy.ssabree.features.board.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.KeywordRepository
import com.ssafy.ssabree.core.repository.PostRepository
import com.ssafy.ssabree.features.board.model.PostUiModel
import com.ssafy.ssabree.features.board.model.toUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class BoardSearchUiState(
    val query: String = "",
    val lastSubmittedQuery: String = "",
    val recentKeywords: List<String> = emptyList(),
    val results: List<PostUiModel> = emptyList(),
    val nextCursor: String? = null,
    val hasNext: Boolean = false,
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val showMinLengthError: Boolean = false,
    val error: String? = null
)

class BoardSearchViewModel(
    private val postRepository: PostRepository,
    private val keywordRepository: KeywordRepository
) : ViewModel() {

    private val pageSize = 20
    private val _uiState = MutableStateFlow(BoardSearchUiState())
    val uiState: StateFlow<BoardSearchUiState> = _uiState.asStateFlow()
    init {
        loadRecentKeywords()
    }

    private fun loadRecentKeywords() {
        val keywords = keywordRepository.getRecentKeywords()
        _uiState.update { it.copy(recentKeywords = keywords) }
    }

    fun onQueryChange(text: String) {
        _uiState.update { state ->
            val trimmed = text.trim()
            if (trimmed == state.lastSubmittedQuery) {
                state.copy(query = text)
            } else {
                state.copy(
                    query = text,
                    results = emptyList(),
                    nextCursor = null,
                    hasNext = false,
                    isLoading = false,
                    isLoadingMore = false,
                    error = null,
                    showMinLengthError = false
                )
            }
        }
    }

    fun onSearchSubmit() {
        val keyword = _uiState.value.query.trim()
        if (keyword.length < 2) {
            _uiState.update { it.copy(showMinLengthError = true) }
            return
        }
        keywordRepository.saveKeyword(keyword)
        loadRecentKeywords()
        viewModelScope.launch {
            load(keyword, cursor = null, append = false)
        }
        _uiState.update { it.copy(lastSubmittedQuery = keyword, showMinLengthError = false) }
    }

    fun onQuickSearch(keyword: String) {
        _uiState.update {
            it.copy(
                query = keyword,
                lastSubmittedQuery = keyword,
                results = emptyList(),
                nextCursor = null,
                hasNext = false,
                isLoading = false,
                isLoadingMore = false,
                error = null,
                showMinLengthError = false
            )
        }
    }

    fun deleteRecentKeyword(keyword: String) {
        keywordRepository.deleteKeyword(keyword)
        loadRecentKeywords()
    }

    fun loadMoreIfNeeded(index: Int) {
        val state = _uiState.value
        val shouldLoadMore = state.hasNext &&
                !state.isLoadingMore &&
                !state.isLoading &&
                index >= state.results.size - 3
        if (!shouldLoadMore) return

        viewModelScope.launch {
            val keyword = state.lastSubmittedQuery
            if (keyword.isEmpty()) return@launch
            load(keyword, state.nextCursor, append = true)
        }
    }

    private suspend fun load(keyword: String, cursor: String?, append: Boolean) {
        if (append) {
            _uiState.update { it.copy(isLoadingMore = true, error = null) }
        } else {
            _uiState.update { it.copy(isLoading = true, error = null) }
        }

        postRepository.getPosts(keyword = keyword, cursor = cursor, limit = pageSize)
            .onSuccess { page ->
                _uiState.update { state ->
                    val newResults = if (append) {
                        state.results + page.posts.map { it.toUiModel() }
                    } else {
                        page.posts.map { it.toUiModel() }
                    }
                    state.copy(
                        results = newResults,
                        nextCursor = page.nextCursor,
                        hasNext = page.hasNext,
                        isLoading = false,
                        isLoadingMore = false,
                        error = null
                    )
                }
            }
            .onFailure { e ->
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isLoadingMore = false,
                        error = e.message
                    )
                }
            }
    }
}
